src/shim.c

```diff
@@ -322,188 +505,133 @@ handle_stdin(struct cc_shim *shim)
+handle_proxy_notification(struct cc_shim *shim, struct frame *fr)
 {
-	uint64_t  seq;
-	char     *buf = NULL;
-	int       outfd;
-	ssize_t   stream_len = 0;
-	ssize_t   ret;
-	ssize_t   offset;
-	int       code = 0;
-
-	if (shim == NULL) {
-		return;
-	}
+	int code = 0;
 
-	buf = read_IO_message(shim, &seq, &stream_len);
-	if ((! buf) || (stream_len <= 0) || (stream_len > HYPERSTART_MAX_RECV_BYTES)) {
-		/*TODO: is exiting here more appropriate, since this denotes
-		 * error communicating with proxy or proxy has exited
-		 */
-		goto out;
-	}
-
-	if (seq == shim->io_seq_no) {
-		outfd = STDOUT_FILENO;
-	} else if (seq == shim->io_seq_no + 1) {//proxy allocates errseq 1 higher
-		outfd = STDERR_FILENO;
-	} else {
-		shim_warning("Seq no %"PRIu64 " received from proxy does not match with\
-				 shim seq %"PRIu64 "\n", seq, shim->io_seq_no);
-		goto out;
+	if ( ! (shim && fr && fr->payload)) {
+		return;
 	}
 
-	if (!shim->exiting && stream_len == STREAM_HEADER_SIZE) {
-		shim->exiting = true;
-		goto out;
-	} else if (shim->exiting && stream_len == (STREAM_HEADER_SIZE+1)) {
-		code = *(buf + STREAM_HEADER_SIZE); 	// hyperstart has sent the exit status
+	if (fr->header.opcode == notif_exitcode) {
+		/* Send disconnect command to proxy and exit
+		 * with the exit code
+		*/
+		code = *(fr->payload);
+		send_proxy_message(shim, type_command, cmd_disconnectshim,
+					 NULL);
 		shim_debug("Exit status for container: %d\n", code);
-		free(buf);
 		restore_terminal();
 		exit(code);
-	}
-
-	/* TODO: what if writing to stdout/err blocks? Add this to the poll loop
-	 * to watch out for EPOLLOUT
-	 */
-	offset = STREAM_HEADER_SIZE;
-	while (offset < stream_len) {
-		ret = write(outfd, buf+offset, (size_t)(stream_len-offset));
-		if (ret <= 0 ) {
-			goto out;
-		}
-		offset += (ssize_t)ret;
-	}
-
-out:
-	if (buf) {
-		free (buf);
+	} else {
+		shim_warning("Unknown notification received from proxy %s: %d\n",
+				shim->proxy_address, fr->header.opcode);
 	}
 }
 
 /*!
- * Handle data on the proxy ctl socket fd
+ * Handle message received from proxy
  *
  *\param shim \ref cc_shim
  */
 void
-handle_proxy_ctl(struct cc_shim *shim)
+handle_proxy_message(struct cc_shim *shim)
 {
-	char buf[LINE_MAX] = { 0 };
-	ssize_t ret;
+	struct frame *fr;
 
-	if (! shim) {
+	fr = read_frame(shim);
+	if ( !fr) {
 		return;
 	}
 
-	ret = read(shim->proxy_sock_fd, buf, LINE_MAX-1);
-	if (ret == -1) {
-		err_exit("Error reading from the proxy ctl socket: %s\n", strerror(errno));
-	} else if (ret == 0) {
-		err_exit("EOF received on proxy ctl socket. Proxy has exited\n");
+	switch (fr->header.type) {
+		case type_response:
+			handle_proxy_response(fr);
+			break;
+		case type_stream:
+			handle_proxy_stream(fr);
+			break;
+		case type_notification:
+			handle_proxy_notification(shim, fr);
+			break;
+		case type_command:
+			shim_warning("Command received from proxy\n");
+			break;
+		default:
+			shim_warning("Unknown frame with type %d received\n",
+				fr->header.type);
```

Comment:
Nice. Again though, might be good to add a log call here to make debugging possible.